freespace2/freespace.cpp

```diff
@@ -1917,6 +1917,9 @@ void game_init()
 
 	animation::ModelAnimationParseHelper::parseTables();
 
+	// Initialize dynamic SEXPs. Must happen before ship init for LuaAI
+	sexp::dynamic_sexp_init();
+
```

Comment:
This looks like a catch-22.  Dynamic sexps need to be initialized after the script system, and the script system needs to be initialized almost last.

Reply:
Not true, actually. The dynamic SEXP system, on parse at least, only really registers the available SEXPs and does not interact with the script system at all. Only three things are important order-wise with dynamic SEXPs and Lua:

1. Lua must not _run_ the on-init hooks before dynamic SEPXs are initialized (this happens later in the function anyways, even before this PR)
2. No dynamic SEXP must be executed (defining them is fine) before the Script system is initialized, as it'll be missing the Lua-state to convert the values
3. No dynamic SEXP must be executed before the Script system has run its on-init hooks

So as far as I can tell, it should be fine to init the dynamic SEXPs before the scripting system, provided you don't try to use one of these SEXPs until scripting is initialized

Reply:
There is this line in `dynamic_sexp_init` which is presumably not valid until the script system is initialized:
```
	Script_system.OnStateDestroy.add(free_lua_sexps);
```

Reply:
I _think_ that that's fine, since it's just adding a listener to when the LuaState object is deallocated, instead of actually requiring access to the LuaState in that moment. And the Script_system object itself already exists, since that has a static lifetime, and only just has no initialized LuaState yet.

Reply:
Ok.

Reply:
I agree with @BMagnu here that this move should be fine. Unfortunately, there is a circular dependency between the two systems since some Lua SEXPs have resources inside the Lua state. The `OnStateDestroy` callback is there to break that dependency so moving the init location should be fine.

However, please also replicate this change in QtFRED to ensure that stays in sync.